File: go/parsers/registered.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Registered struct {
	gen *gen.Registered
}

func NewRegisteredFields(g *gen.Registered) messages.RegisteredFields {
	return &Registered{gen: g}
}

func (r *Registered) RequestID() uint64 {
	return r.gen.RequestID()
}

func (r *Registered) RegistrationID() uint64 {
	return r.gen.RegistrationID()
}

func RegisteredToCapnproto(m *messages.Registered) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	registered, err := gen.NewRootRegistered(seg)
	if err != nil {
		return nil, err
	}

	registered.SetRequestID(m.RequestID())
	registered.SetRegistrationID(m.RegistrationID())

	data, err := msg.Marshal()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeRegistered, data, nil), nil
}

func CapnprotoToRegistered(data []byte) (*messages.Registered, error) {
	msg, err := capnp.Unmarshal(data)
	if err != nil {
		return nil, err
	}

	registered, err := gen.ReadRootRegistered(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewRegisteredWithFields(NewRegisteredFields(&registered)), nil
}

File: python/wamp_msgs_capnp/parsers/registered.py
import os
from pathlib import Path

import capnp
from wampproto.messages.registered import Registered, IRegisteredFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
schema_file = os.path.join(root_dir, "schemas", "registered.capnp")
registered_capnp = capnp.load(str(schema_file))


class RegisteredFields(IRegisteredFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def registration_id(self) -> int:
        return self._gen.registrationID


def registered_to_capnproto(r: Registered) -> bytes:
    registered = registered_capnp.Registered.new_message()
    registered.requestID = r.request_id
    registered.registrationID = r.registration_id

    packed_data = registered.to_bytes_packed()
    return helpers.prepend_header(Registered.TYPE, packed_data, b"")


def capnproto_to_registered(data: bytes) -> Registered:
    message_data, _ = helpers.extract_message(data)
    registered_obj = registered_capnp.Registered.from_bytes_packed(message_data)

    return Registered(RegisteredFields(registered_obj))
